lib/speed/time-collector.js

```diff
@@ -0,0 +1,185 @@
+    const sumOfSquares = testBuffer.reduce((sum, time) => {
+      return Math.pow((time - averageTime), 2) + sum;
+    }, 0);
+    const delta = Math.sqrt(sumOfSquares) * 100 / bufferSize;
+    let totalTime = 0;
+    let anomalyTime = 0;
+    let anomalyCount = 0;
+    testBuffer.forEach(time => {
+      if (Math.abs(time - averageTime) > delta) {
+        anomalyCount++;
+        anomalyTime += time;
+      } else {
+        totalTime += time;
+      }
+    });
+    const percent = anomalyCount / bufferSize * 100;
+    if (percent <= this.anomalyPercent && anomalyCount) {
+      fnEnvironment.anomalyCount += anomalyCount;
+      fnEnvironment.anomalyTime += anomalyTime;
+      totalTime += anomalyCount * averageTime;
+    } else {
+      totalTime += anomalyTime;
+    }
+    fnEnvironment.buffer = [];
+    fnEnvironment.totalTime += totalTime;
+  }
+
+  getStatus(
+    // Function that get optimization status of test function
+    fn //Function, test function
+    // Returns: String, optimization status
+  ) {
+    const OPT_BITS = [
+      /*  1 */ 'function',
+      /*  2 */ 'never',
+      /*  4 */ 'always',
+      /*  8 */ 'maybe',
+      /* 16 */ 'opt',
+      /* 32 */ 'turbofan',
+      /* 64 */ 'interpreted'
+    ];
+    const optStatus = %GetOptimizationStatus(fn); // eslint-disable-line
+    const results = [];
+    OPT_BITS.forEach((name, n) => {
+      if (n === 0) return;
+      if (Math.pow(2, n) & optStatus) results.push(name);
+    });
+    return  results.length ? results.join(',') : '---';
+  }
+
+  sync(
+    // Function that measures the time of synchronous functions
+    fn, // Function, synchronous test function
+    fnEnvironment // Object, object with auxiliary values for testing
+  ) {
+    for (let i = 0; i < this.startCount; i++) {
+      %OptimizeFunctionOnNextCall(fn); // eslint-disable-line
+      this.intermediateValue.add(fn(), i);
+    }
+    for (let i = 0; i < this.count; i++) {
+      const begin = process.hrtime();
+      const result = fn();
+      const end = process.hrtime(begin);
+      fnEnvironment.add(end[0] * 1e9 + end[1], i);
+      this.intermediateValue.add(i, result);
+    }
+    fnEnvironment.optimizationStatus = this.getStatus(fn);
+    this.makeResults(fnEnvironment);
+  }
+
+  async(
+    // Function that measures the time of asynchronous functions
+    fn, // Function, asynchronous test function (callback)
+    fnEnvironment // Object, object with auxiliary values for testing
+  ) {
+    const testFunction = (begin, i) => fn((err, result) => {
+      const end = process.hrtime(begin);
+      this.intermediateValue.add(i, err, result);
+      fnEnvironment.add(end[0] * 1e9 + end[1], i);
+      if (++i < this.count) {
+        testFunction(process.hrtime(), i);
+        return;
+      }
+      fnEnvironment.optimizationStatus = this.getStatus();
+      this.makeResults(fnEnvironment);
+    });
+
+    const optimiseFunction = i => fn((err, result) => {
```

Comment:
What id `i` ? Bad naming for argument. Why do you use recursion here?

Reply:
I want to start testing an asynchronous function only after the previous one is executed function. I call the following function from the callback of the previous function.